DANS-KNAW/dataverse
# Configurable account migration (test PR)

**What this PR does / why we need it**: Currently, when a user logs in to a migrated builtin user account for which the password is encrypted with a legacy hash algorithm (e.g., SHA-1), Dataverse will display a password reset page that also requires the user to accept the Terms of Use again. After a new password is provided it gets encrypted with the latest and greatest algorithm.

There are several limitations and issues with the current reset page:
* The introductory message talks about an upgrade from Dataverse 4.0, which is not necessarily the case.
* There is no way to only re-encrypt the current user password, i.e. the user is forced to provide a new one. However, for example DANS would like to just require the user to accept the Terms of Use again and no change the password.

Example of the current screen:
![image](https://user-images.githubusercontent.com/3862694/120212984-fe90fc00-c232-11eb-9e13-0cfc115100c5.png)

**Which issue(s) this PR closes**: N/A

**Special notes for your reviewer**: This PR introduces the following database settings:
* `:SilentPasswordAlgorithmUpdateEnabled`: if set to `true` with a `PUT` request, the "New Password" and "Retype Password" entry fields are hidden and the existing password is automatically re-encrypted on clicking the continue button.
* `:PVCustomPasswordResetAlertMessage`: the text that should appear in the blue banner at the top of the passwordreset page. 
* `:CustomPasswordResetButton`: the caption of the button to continue. (If `:SilentPasswordAlgorithmUpdateEnabled` is enabled you should override the default text which is "Reset Password".)
* `:CustomPasswordResetAlertIntro`: the introductory text.

Note that, if the password turns out, not to comply with the minimal security requirements, the user is presented with the regular password reset page on clicking Continue, and will be required to provide a new password.

**Suggestions on how to test this**:

1. Configure the Dataverse instance to automatically reencrypt password, e.g.:
   ```bash
    curl -X PUT -d "true" http://localhost:8080/api/admin/settings/:SilentPasswordAlgorithmUpdateEnabled
    curl -X PUT -d '{0} Our Terms of Use have been updated:{1} Please check the box below and click Continue' \
         http://localhost:8080/api/admin/settings/:PVCustomPasswordResetAlertMessage
    curl -X PUT -d "Continue" http://localhost:8080/api/admin/settings/:CustomPasswordResetButton
    curl -X PUT -d 'Please accept our new Terms of Use.  Also, note that you can connect your account to your institutional account, or ORCID, Google or GitHub account. To do so, log out after completing this step. Log in again with one of “Log in with our institutional account”, “ORCID”, “Google” or “GitHub”. ' http://localhost:8080/api/admin/settings/:CustomPasswordResetAlertIntro
    ```
2. Create a SHA-1 password, e.g.,  `echo -n mypassword001 | openssl dgst -binary -sha1 | openssl base64` 
3. Update the `builtinuser` table: 
    ```sql
    UPDATE builtinuser 
    SET encryptedpassword = '<result of previous command>'
    WHERE username = '<user to edit>';

    UPDATE builtinuser 
    SET passwordencryptionversion = 0
    WHERE username = '<user to edit>';
    ```
4. Log in. You should see a page similar to the one below:
    ![image](https://user-images.githubusercontent.com/3862694/120214948-73fdcc00-c235-11eb-87bf-3d4dbf577eb2.png)

(Note: you can customize the build number with the script `dataverse/scripts/custom-build-number` in the Dataverse code.)

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**: See previous item.

**Is there a release notes update needed for this change?**: No

**Additional documentation**:
